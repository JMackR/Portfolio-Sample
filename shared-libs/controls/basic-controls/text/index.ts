export * from './text'
export * from './text.props.base'
export * from './text.props'
