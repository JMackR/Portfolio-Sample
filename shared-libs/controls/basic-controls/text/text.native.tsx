import React from 'react'
import { Text as RNText, TextStyle } from 'react-native'
import { useColorForTextColor, useFontForTextType } from '@sample/themes/hooks'
import { shadow } from '@sample/themes/shadow'
import { TextProps } from './text.props.native'

const autoLineHeight = {
  lineHeight: undefined,
} as TextStyle

export const Text: React.FunctionComponent<TextProps> = (props) => {
  const {
    numberOfLines,
    children,
    text,
    textType,
    textAlign,
    textDecorationLine,
    onPress,
    testID,
    selectable,
    color: textColor,
    dropShadow,
    ellipsizeMode,
  } = props
  const font = useFontForTextType(textType || 'primary3')
  const color = useColorForTextColor(textColor || 'primary1')
  const textShadow = dropShadow ? shadow.shadow : undefined

  const content = text || children
  return (
    <RNText
      style={[font, { color, textAlign, textDecorationLine }, textShadow]}
      numberOfLines={numberOfLines}
      ellipsizeMode={ellipsizeMode}
      onPress={onPress}
      selectable={selectable}
      testID={testID || 'text'}
      accessibilityRole={'text'}
      // accessibilityLabel={testID || 'text'}
    >
      {content}
    </RNText>
  )
}
