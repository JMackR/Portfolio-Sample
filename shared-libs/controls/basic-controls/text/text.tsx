import React, { forwardRef, useImperativeHandle, useRef, MutableRefObject } from 'react'
import { useFontForTextType, useColorForTextColor } from '@sample/themes'
import { convertFontToWebStyleSheet } from '@sample/themes/utility/font-conversions'
import { StyleSheet, css, StyleDeclaration } from 'aphrodite/no-important'
import { TextProps, TextRef } from './text.props'

/**
 * This is a simple text component.  It includes all behavior displayed by text in the sircles app.
 * @param props see TextProps for more info.
 */

export const Text = forwardRef<TextRef, TextProps>((props, ref) => {
  const {
    numberOfLines,
    children,
    textType,
    text,
    textAlign,
    textDecorationLine,
    onPress,
    color,
    whiteSpace = 'inherit',
  } = props
  const content = text || children
  let textStyle: StyleDeclaration = {
    whiteSpace,
    textAlign,
    textDecorationLine,
  }
  if (color === 'primary1') {
    textStyle = {
      ...textStyle,
      ':hover': {
        cursor: 'pointer',
      },
    }
  }
  if (numberOfLines) {
    textStyle = {
      ...textStyle,
      display: '-webkit-box',
      WebkitBoxOrient: 'vertical',
      WebkitLineClamp: numberOfLines,
      overflow: 'hidden',
      textOverflow: 'ellipsis',
    }
  }
  const STYLES = StyleSheet.create({
    textStyle,
  })
  const font = useFontForTextType(textType || 'primary3')
  const fontColor = useColorForTextColor(color || 'primary1')
  const fontStyles = convertFontToWebStyleSheet(font, fontColor, 'inline')

  const contentRef = useRef() as MutableRefObject<any>

  useImperativeHandle(ref, () => ({
    isContentTruncated: () => {
      if (contentRef.current) {
        return contentRef.current.scrollHeight > contentRef.current.offsetHeight
      }
      return false
    },
  }))

  return (
    <div className={css(fontStyles.text, STYLES.textStyle)} onClick={onPress} ref={contentRef}>
      {content}
    </div>
  )
})
