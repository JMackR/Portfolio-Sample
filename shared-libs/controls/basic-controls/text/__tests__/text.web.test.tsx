import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { StyleSheetTestUtils } from 'aphrodite/no-important'
import { createSerializer } from 'jest-aphrodite-react/no-important'
import renderer from 'react-test-renderer'
import { Text } from '../text'
import { TextProps } from '../text.props'

expect.addSnapshotSerializer(createSerializer(() => StyleSheetTestUtils, { removeVendorPrefixes: true }))

describe('Text', () => {
  let props: TextProps

  beforeEach(() => {
    props = {
      text: 'Test Text',
    }
  })

  it('should render the text', () => {
    const { getByText } = render(<Text {...props} />)

    expect(getByText(props.text as string)).toBeDefined()
  })

  describe('children are passed in', () => {
    describe('there is still text', () => {
      it('should render the text', () => {
        const { getByText } = render(<Text {...props}>Children</Text>)

        expect(getByText(props.text as string)).toBeDefined()
      })
    })

    describe('there is still text', () => {
      beforeEach(() => {
        props.text = undefined
      })

      it('should render the children', () => {
        const { getByText } = render(<Text {...props}>Children</Text>)

        expect(getByText('Children')).toBeDefined()
      })
    })
  })

  describe('the user clicks the text', () => {
    beforeEach(() => {
      props.onPress = jest.fn()
    })

    it('should call onPress', () => {
      const { getByText } = render(<Text {...props} />)

      const text = getByText(props.text as string)
      fireEvent.click(text)

      expect(props.onPress).toHaveBeenCalledTimes(1)
    })
  })
})

describe('Text Snapshot Tests', () => {
  let props: TextProps
  beforeEach(() => {
    props = {
      text: 'Test Text',
    }
  })

  it('text creates properly', () => {
    const tree = renderer.create(<Text {...props} />)

    expect(tree).toMatchSnapshot()
  })

  describe('color is primary1', () => {
    beforeEach(() => {
      props.color = 'primary1'
    })

    it('text creates properly', () => {
      const tree = renderer.create(<Text {...props} />)

      expect(tree).toMatchSnapshot()
    })
  })

  describe('number of lines is passed in', () => {
    beforeEach(() => {
      props.numberOfLines = 2
    })

    it('text creates properly', () => {
      const tree = renderer.create(<Text {...props} />)

      expect(tree).toMatchSnapshot()
    })
  })
})
