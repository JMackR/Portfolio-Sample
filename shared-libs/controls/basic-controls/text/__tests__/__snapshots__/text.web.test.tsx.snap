// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Snapshot Tests color is primary1 text creates properly 1`] = `
.text-textStyle {
  color: #696969;
  font-size: 16px;
  font-weight: 700;
  font-family: ABCSocial-Bold, sans-serif;
  display: inline;
  white-space: inherit;
  text-align: undefined;
  text-decoration-line: undefined;
}

.text-textStyle:hover {
  cursor: pointer;
}

<div
  className="text-textStyle"
>
  Test Text
</div>
`;

exports[`Text Snapshot Tests number of lines is passed in text creates properly 1`] = `
.text-textStyle {
  color: #696969;
  font-size: 16px;
  font-weight: 700;
  font-family: ABCSocial-Bold, sans-serif;
  white-space: inherit;
  text-align: undefined;
  text-decoration-line: undefined;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div
  className="text-textStyle"
>
  Test Text
</div>
`;

exports[`Text Snapshot Tests text creates properly 1`] = `
.text-textStyle {
  color: #696969;
  font-size: 16px;
  font-weight: 700;
  font-family: ABCSocial-Bold, sans-serif;
  display: inline;
  white-space: inherit;
  text-align: undefined;
  text-decoration-line: undefined;
}

<div
  className="text-textStyle"
>
  Test Text
</div>
`;
